software/controller/lib/debug/commands.h

```diff
@@ -0,0 +1,219 @@
+    // max or the buffer is empty.
+
+    uint32_t count;
+    for (count = 0; count < context->max_response_length; count++) {
+      std::optional<uint8_t> next_char = debug.GetNextCharFromBuffer();
+      if (next_char == std::nullopt)
+        break;
+      context->response[count] = *next_char;
+    }
+
+    context->response_length = count;
+    return ErrorCode::kNone;
+  }
+};
+
+// Peek command.
+// Allows us to read raw memory values.
+//
+// The data passed to the command consists of a 32-bit starting address and
+// a 16-bit byte count.
+class PeekHandler : public Handler {
+public:
+  PeekHandler() = default;
+  ErrorCode Process(Context *context) override {
+    // The data passed to this command consists of a 4 byte address
+    // and a two byte count of how many bytes to return
+    // Length should be exactly 6, but if more data is passed
+    // I'll just ignore it.
+    if (context->request_length < 6)
+      return ErrorCode::kMissingData;
+
+    uint32_t address = u8_to_u32(&context->request[0]);
+    uint32_t count = u8_to_u16(&context->request[4]);
+
+    // Limit the number of output bytes based on buffer size
+    if (count > context->max_response_length)
+      count = context->max_response_length;
+
+    // Some registers can't handle byte accesses, so rather then just
+    // use a simple memcpy here I will do 32-bit or 16-bit accesses
+    // as long as both the address and count are aligned to 32 or 16 bits.
+    if (!(address & 3) && !(count & 3)) {
+      uint32_t *ptr = reinterpret_cast<uint32_t *>(address);
+
+      for (int byte_number = 0; byte_number < count / 4; byte_number++) {
+        uint32_t value = *ptr++;
+        u32_to_u8(value, &context->response[4 * byte_number]);
+      }
+    } else if (!(address & 1) && !(count & 1)) {
+      uint16_t *ptr = reinterpret_cast<uint16_t *>(address);
+      for (int byte_number = 0; byte_number < count / 2; byte_number++) {
+        uint16_t value = *ptr++;
+        u16_to_u8(value, &context->response[2 * byte_number]);
+      }
+    } else {
+      memcpy(context->response, reinterpret_cast<void *>(address), count);
+    }
+
+    context->response_length = count;
+    return ErrorCode::kNone;
+  }
+};
+
+// Poke command.
+// Allows us to write raw values in memory.  Use with caution!
+// The data passed to the command consists of a 32-bit starting address and one
+// or more data bytes to be written to consecutive addresses starting at the
+// given address
+class PokeHandler : public Handler {
+public:
+  PokeHandler() = default;
+  ErrorCode Process(Context *context) override {
+
+    // Total command length must be at least 5.  That's
+    // four for the address and at least one data byte.
+    if (context->request_length < 5)
+      return ErrorCode::kMissingData;
+
+    uint32_t address = u8_to_u32(&context->request[0]);
+
+    uint32_t count = context->request_length - 4;
+
+    // If both the address and count are multiples of 4, I write
```

Comment:
This is pretty clever. Can it be tested? Also, sorry if I am being naive but is this completely new code or was this migrated here from another file?

Reply:
Moved to a common library (`binary_utils`)